Generalized SMC Algorithms

philipwosull opened the pull request:
Adding new sampling spaces, splitting methods, and smc weights among other things

philipwosull committed:
tweaked find edge to cut function to better integrate with merge split

philipwosull committed:
more tweaks to the updating functions from a cut tree

philipwosull committed:
started working on merge split stuff

philipwosull committed:
built out skeleton for adding merge split steps

philipwosull committed:
almost done merge split. Just need to add MH ratio rejection step

philipwosull committed:
mh rejection step now added for generalized split still need to test district only

philipwosull committed:
merge split now works for one district split as well

philipwosull committed:
fixed issue with not correctly spacing out merge split steps

philipwosull committed:
fixed issue with bad mcmc weights. Now just leaves them unchanged

philipwosull committed:
added support for counting specific region ids within a plan instead of all of them

philipwosull committed:
tweaked merge split to just do 1/acceptance rate again, will make it variable parameter in future

philipwosull committed:
removed reference to file not on github

philipwosull committed:
made original smc support not multiprocessing

philipwosull committed:
makes it so when not doing multiprocess the output from each run is printed

philipwosull committed:
made it possible to do a final merge split step at the end

philipwosull committed:
changed so its now possible to run merge split after the final smc step

philipwosull committed:
can now specify multple of expected mcmc steps

philipwosull committed:
Cory redist merge split now works

philipwosull committed:
attempting to rename package to gredist

philipwosull committed:
two more files

CoryMcCartan committed:
rename types

philipwosull committed:
fixed final compilation issue. package successfully renamed

philipwosull committed:
started major code refactoring. Most things returned from function are now matrices

philipwosull committed:
added function to reorder by oldest split in cpp

philipwosull committed:
significant refactor. Now outputs integer types back to R as integers so uses about half the space as before

philipwosull committed:
signficantly sped up weight computation by using hash map and doing everything in single pass through

philipwosull committed:
added support for simple backwards kernel weights

philipwosull committed:
added support for splitting regions where one of the pieces is within a range. Also started working on manual splitting stuff

philipwosull committed:
added function for complete manual splitting of map and saving the forest and cut trees

philipwosull committed:
renamed redist gsmc file

philipwosull committed:
last push before refactor to supporting multiple plan subtypes

philipwosull committed:
rewrote functions to support abstract plan type

philipwosull committed:
switched sampling in smc back to local random.h file, not cpp random distribs

philipwosull committed:
fixed k estimation for gsmc

philipwosull committed:
added support for two new splitting types and verified validation still works for gsmc plans

philipwosull committed:
added support for running with non-blank initial plans and only running partially

philipwosull committed:
added forest adjacency stuff

philipwosull committed:
updated tree splitter to avoid unnecssary vector allocations for pop_below and parents

philipwosull committed:
passes basic validation:
git push

philipwosull committed:
spanning forest passes validation for uniform valid edge and expo weighted best edge

philipwosull committed:
added new splitter types

philipwosull committed:
validation still works after minor code refactoring

philipwosull committed:
started working on adding proper weights for custom size splitting schedule

philipwosull committed:
weights now partially work for custom sizes but still slightly off

philipwosull committed:
some minor cleanup of tree splitting functions

philipwosull committed:
tweaked k estimation but half splits and by two still don't work

philipwosull committed:
tweaked R files so now one function handles gsmc cpp call and other two are wrappers

philipwosull committed:
tweaked weight computation, still working on making custom size scheudle more correct

philipwosull committed:
made seperate function for getting ajd regions and effective forest boundary len

philipwosull committed:
made abstract splitting schedule base class and derived classes for any size and district only. Verified they work on 6x6 validation

philipwosull committed:
starting working on constraint abstract class and scoring function class to handle constraints

philipwosull committed:
optimal weight function now has support for scoring function and streamlined code to work for graph and forest space

philipwosull committed:
now supports custom adj sizing

philipwosull committed:
fixed hinge constraint bug

philipwosull committed:
tweaked original smc to allow for setting processes and num threads per process seperately

philipwosull committed:
tweaked region selection prob in gsmc to now be size^alpha

philipwosull committed:
RNG generation is now threadsafe

philipwosull committed:
tweaked diagnostic that redist_gsmc puts out

philipwosull committed:
consolidated mergesplit files and changed to just specify desired number of samples, not total steps

philipwosull committed:
make PRNG more robust with long_jump functionality

philipwosull committed:
Adding Plan method for computing log tau terms

philipwosull committed:
cleaned up code and removed extraneous plan variables

philipwosull committed:
added compactness parameter and all alg versions passed 6x6 validation

philipwosull committed:
removed more unncessary attributes from plan object

philipwosull committed:
removed recursion from functions which use cut tree to update plans

philipwosull committed:
refactor of get valid edges in tree with speed up by only requiring one pass through

philipwosull committed:
Now removed unnessecary copying in the splitting step

philipwosull committed:
removed duplicate tree splitters, now only uses one for entire gsmc function

philipwosull committed:
forgot new UST Sampler files

philipwosull committed:
some minor tweaks to starting prinout message

philipwosull committed:
removed feature unnesarily computing log tau for simple graph space weights

philipwosull committed:
added parallel versions of some redistmetrics functions to speed up summary stat calculation for huge plans

philipwosull committed:
added more parallel versions of some redistmetrics functions

philipwosull committed:
adding weights file

philipwosull committed:
more tweaks to make parallel redistmetric functions more thread safe. Still don't know why parallel_splits seg faults sometimes

philipwosull committed:
fixed issue with parallel polsby popper. If no pool object RcppThread ParallelFor keeps static variables in global R memory. avoid in future

philipwosull committed:
implemented new merge split using Plan class. Passes validation

philipwosull committed:
some more code refactor to help support linking edge stuff

philipwosull committed:
skeleton of linking edge done. Just need to add multigraph tau

philipwosull committed:
linking edge space sampling passes validation

philipwosull committed:
some code tweaks

philipwosull committed:
minor typo in tree splitter types file

philipwosull committed:
exposed exported generic gsmc file as export

philipwosull committed:
forgot generic redist man page

philipwosull committed:
more man page tweaks

philipwosull committed:
removed unneccesary copy in smc plans function

philipwosull committed:
reduced memory usage in plan reorder

philipwosull committed:
smc merge split working for forest, still doesn't pass validation for linking edge

philipwosull committed:
caught error with not correctly updating linking probabilities/computing them initiallly

philipwosull committed:
smc+ms works for linking edge

philipwosull committed:
refactored diagnostics and added code to track the size of regions we draw trees on

philipwosull committed:
added for mcmc too

philipwosull committed:
added constants for alg types

philipwosull committed:
adding county check in merge split

philipwosull committed:
added county split checking

philipwosull committed:
added more diagnostics for merge split seperation bug

philipwosull committed:
hopefully fixed issue with incorrect hiearchical tree drawing for generalized region splits

philipwosull committed:
trying to turn off pop bounds in sample_sub_ust to see if that fixes issue

philipwosull committed:
trying again to fix

philipwosull committed:
indexing typo in update

philipwosull committed:
added quick multithreaded sorting function for district statistics

philipwosull committed:
had non uploaded function in Rcpp file

philipwosull committed:
didnt remove stuff from Rcpp exports last time

philipwosull committed:
very hacky fix for county issue GRAPH SPACE ONLY

philipwosull committed:
adding functions for computing log optimal weights and target density

philipwosull committed:
another bug in county checking, quick fix but needs more verification

philipwosull committed:
attempted to fix county boundary counting issue

philipwosull committed:
major refactor. Switched from arma columns to uint8 vectors for plan ids

philipwosull committed:
starting to refactor forests to be smaller

philipwosull committed:
tweaking forest assignment to take more space upfront

philipwosull committed:
Now support 65535 counties instead of 255

philipwosull committed:
hopefully counties now work for graph plans optimal weights

philipwosull committed:
fixed another uint issue

philipwosull committed:
switched determinants to PSD version and added shallow copy

philipwosull committed:
refactor to avoid using as much memory when no intial plans

philipwosull committed:
more memory usage refactoring

philipwosull committed:
more memory usage reduction

philipwosull committed:
resampling done in c++ to save memory

philipwosull committed:
more stuff done in Rcpp to save mem

philipwosull committed:
made extra printing controlled by verbose

philipwosull committed:
fixed bug with incorrect copying for odd split numbers

philipwosull committed:
fixed minor typo

philipwosull committed:
added functions to compute log target density and the log target for each region

philipwosull committed:
added incomplete implementation of pair hash

philipwosull committed:
overhauled county component stuff for graph and forest space

philipwosull committed:
smc and smc+mcmc pass validation now

philipwosull committed:
constraint code now scores districts only on per constraint basis. Default for most to is to score all regions:

philipwosull committed:
mergesplit tweak

philipwosull committed:
added support for MMD plans and nseats for partial SMD plans

philipwosull committed:
Added but not tested code for multi member district district only splits

philipwosull committed:
added hard constraints to manual weights cpp code

philipwosull committed:
caught potential bug in MMD split district only where big districts could be split

philipwosull committed:
nvm allowing districts to be split again

philipwosull committed:
renamed and changed redist_gsmc function inputs to match old redist_smc

philipwosull committed:
some more cleanup tweaks to gsmc cpp file

philipwosull committed:
added support for seeding random trees on initialized non blank forest plans

philipwosull committed:
Major refactor of code, eliminated county components and replaced with plan multigraph object used for finding adj pairs

philipwosull committed:
fixed minor bug in simple weights

philipwosull committed:
caught bug in not catching diff adjacent components

philipwosull committed:
fixed error in redist_gsmc where it would not do split districts only

philipwosull committed:
change multigraph to be vec of vec of int array of size 3

philipwosull committed:
now display the number of threads in the threadpool

philipwosull committed:
renamed redist gsmc to redist smc and moved old smc to legacy version. will delete legacy in future

philipwosull committed:
renamed everything from gredist back to redist

philipwosull committed:
more renaming back to redist stuff and reset plot_maps file

philipwosull committed:
attempting to add seq_alpha back but failing validation

philipwosull committed:
seq alpha passes validation

philipwosull committed:
removed old smc cpp code and replaced it with new code

philipwosull committed:
fixed bug in MMD district splits where incorrect split sizes from previous steps weren't being cleared

philipwosull committed:
starting to overhaul diagnostics

philipwosull committed:
made it so you always compute multidsistrict selection probability for MMD just in case

philipwosull committed:
more tweaks and fixed a ms bug with tree size counts

philipwosull committed:
removed some pointless tests and fixed issue with null nseats for reference plans in redist ms

philipwosull committed:
removed so non used functions, consolidated a file, and reskinned mergesplit k estimation function

philipwosull committed:
caught issue with multigraph check if hier valid not working properly

philipwosull committed:
added incumbency constraint to updated smc code, avoided duplicate by using templating

philipwosull committed:
more constraints added to new code

philipwosull committed:
overhauled Wilson code to reuse as many vectors as possible

philipwosull committed:
made tree pop stacks fixed size

philipwosull committed:
updating SMC tests for new code interface changes

philipwosull committed:
updated merge split tests for new function inputs

philipwosull committed:
added new soft and hard plan constraints and other tweaks

philipwosull committed:
code now passes all tests on Philip mac

philipwosull committed:
removed all parallel redistmetrics functions. Moved to gsmcs repo

philipwosull committed:
clean ups

philipwosull committed:
fixing merge conflicts

philipwosull committed:
updated code to avoid unnecessary repeated memory allocation in weights calculation

philipwosull committed:
Added updated version of eval_segregation and eval_grp_pow and grp_pow, compet, and segregation constraints to updated code

CoryMcCartan commented on R/diagnostics.R:
```diff
@@ -320,3 +463,106 @@ diag_rhat <- function(x, grp, split = FALSE) {
+#' @export
+get_k_step_ancestors <- function(parent_mat, steps_back = NULL, start_col = NULL){
+    # check the matrix is not zero indexed
+    assertthat::assert_that(
```
We don't import `assertthat`

Please convert to `if` statements that throw `cli_abort` errors if violated

philipwosull replied:
Control F and removed all uses of `assertthat`.

CoryMcCartan commented on R/diagnostics.R:
```diff
@@ -320,3 +463,106 @@ diag_rhat <- function(x, grp, split = FALSE) {
+#' @returns indices of originals ancestors for particles at every iteration. So
+#' entry [s,j] is the original ancestor of particle j after step s
+#' @md
+#' @export
```
why is this exported?

philipwosull replied:
No longer exported

CoryMcCartan commented on R/diagnostics.R:
```diff
@@ -320,3 +463,106 @@ diag_rhat <- function(x, grp, split = FALSE) {
+    for (j in start_col:(start_col - steps_back + 1)) {
+        ancestor <- parent_mat[ancestor, j]
+    }
+    return(ancestor)
```
no trailing `return()` needed

CoryMcCartan commented on R/diagnostics.R:
```diff
@@ -320,3 +463,106 @@ diag_rhat <- function(x, grp, split = FALSE) {
+        original_ancestor_mat
+    )
+
+    return(original_ancestor_mat)
```
no trailing `return()` needed
just have the last line be `cbind(...)`

CoryMcCartan commented on R/plan_calculations.R:
What do these functions do and why do we need them?

We don't want to expose this stuff to users generally, IMO

CoryMcCartan commented on R/plans_helpers.R:
```diff
@@ -50,30 +50,34 @@ check_tidy_types <- function(map, .data) {
 #' @concept analyze
 #' @export
-tally_var <- function(map, x, .data = pl()) {
+tally_var <- function(map, x, .data = pl(), num_threads = 1) {
```
change to `ncores` to be consistent

CoryMcCartan commented on R/plans_helpers.R:
```diff
@@ -50,30 +50,34 @@ check_tidy_types <- function(map, .data) {
 #' @export
 group_frac <- function(map, group_pop, total_pop = map[[attr(map, "pop_col")]],
-                       .data = pl()) {
+                       .data = pl(), num_threads = 1) {
```
`ncores`

CoryMcCartan commented on R/plans_helpers.R:
```diff
@@ -121,18 +125,25 @@ avg_by_prec <- function(plans, x, draws = NA) {
 #' @concept analyze
 #' @export
-plan_parity <- function(map, .data = pl(), ...) {
+plan_parity <- function(map, .data = pl(), num_threads = 1, rep_by_district = TRUE) {
```
why do we need `rep_by_district`? We always do that

`ncores` too

philipwosull replied:
Changed for high performance version when there's a large number of plans. Just added that to the paper repo and reverted this by to duplicating for each district.

CoryMcCartan commented on R/redist_alg_helpers.R:
These are internal fns to share validation across the various algorithm/sampling functions?

philipwosull replied:
Yes, shared input validation functions. Removed old custom splitting validation code.

CoryMcCartan commented on R/redist_alg_helpers.R:
```diff
@@ -0,0 +1,643 @@
+    }
+
+
+    return(list(
```
no trailing `return()`

CoryMcCartan commented on R/redist_alg_helpers.R:
```diff
@@ -0,0 +1,643 @@
+                 with {.arg compactness == 1). This may disrupt efficient sampling.")
+    }
+    constraints <- as.list(constraints) # drop data attribute
+    return(constraints)
```
no trailing `return()`

I'll stop adding comments about this but just ctrl-F for instances of this and remove

CoryMcCartan commented on R/redist_alg_helpers.R:
```diff
@@ -0,0 +1,643 @@
+        if(!"estimate_cut_k" %in% names(splitting_params)){
+            # If no adapt k mentioned default is to estimate
+            splitting_params$estimate_cut_k <- T
+            #cli::cli_abort("For Naive Top K splitting method {.arg estimate_cut_k} must be included in {.arg splitting_params}")
```
remove commented out lines

make sure code is wrapped to 80 chars

CoryMcCartan commented on R/redist_alg_helpers.R:
```diff
@@ -0,0 +1,643 @@
+}
+
+
+# TODO: Add flag for split district only and if so make sure that
```
Handle this TODO and/or remove this comment

CoryMcCartan commented on R/redist_alg_helpers.R:
```diff
@@ -0,0 +1,643 @@
+validate_initial_region_id_mat <- function(init_region_ids_mat, V, nsims, init_num_regions){
+    # check that matrix dimension is V by nsims
+    if(!is.matrix(init_region_ids_mat)){
+        cli::cli_abort("{.arg init_region_ids_mat} must be a matrix.")
```
This whole file: no need for `cli::` prefix because we import this fn

CoryMcCartan commented on R/redist_alg_helpers.R:
```diff
@@ -0,0 +1,643 @@
+
+
+
+# TODO: Make this work with partial splits. Right now only works with
```
handle or remove

CoryMcCartan commented on R/redist_alg_helpers.R:
```diff
@@ -0,0 +1,643 @@
+#' for each split.
+#'
+#' @returns A counties label vector
+OLD_validate_cut_sizes <- function(ndists, num_splits,
```
do we use this function? if so, rename; if not, remove

CoryMcCartan commented on R/redist_alg_helpers.R:
```diff
@@ -0,0 +1,643 @@
+    }
+
+    # for now just return. No other validation performed
+    return()
```
`return(NULL)` for clarity

then delete remaining code in this fn

CoryMcCartan commented on R/redist_constants.R:
```diff
@@ -0,0 +1,31 @@
+
+
+## Forward Kernel Splitting types
+NAIVE_K_SPLITTING <- "top_k" # pick top k unif at random
```
not a fan of the one-off constants... can we just find-replace all of these with `top_k`?

the vectors are OK though. Would prefer to move these to the top of `redist_smc.R` however

philipwosull replied:
These are used in at least 4 files (`redist_smc.R`, `redist_ms.R`, `diagnostics.R`, and `redist_alg_helpers.R`). I think its more convenient/easier to change if needed then just using the raw string.

CoryMcCartan commented on R/redist_constants.R:
```diff
@@ -0,0 +1,31 @@
+
+
+# Algorithm Types
+SMC_ALG_TYPE <- "smc"
```
Eliminate these

CoryMcCartan commented on R/redist_constr.R:
```diff
@@ -210,13 +210,48 @@ add_to_constr <- function(constr, name, new_constr) {
 #'
 #' @param constr A [redist_constr()] object
 #' @param strength The strength of the constraint. Higher values mean a more restrictive constraint.
+#' @param score_districts_only Whether or not to apply the constraints to
```
Do we really need this argument? And why does it default to `TRUE`?

CoryMcCartan commented on R/redist_constr.R:
```diff
@@ -655,10 +721,110 @@ add_constr_custom <- function(constr, strength, fn) {
+#' @param fn A function
+#' @rdname constraints
+#' @export
+add_hard_constr_custom_plan <- function(constr, fn) {
```
See Slack discussion. Don't think we should separate pre- vs post- constraints. Instead, add a `thresh=Inf` arg

CoryMcCartan commented on R/redist_map.R:
```diff
@@ -9,8 +9,10 @@
-new_redist_map <- function(data, adj, ndists, pop_bounds, pop_col = "pop",
-                           adj_col = "adj", add_adj = TRUE, existing_col = NULL) {
+new_redist_map <- function(
+        data, adj, ndists, total_seats, district_seat_sizes,
```
note for @christopherkenny : this is a breaking change in arg order but I'm actually OK with it because it's weird to add additional seats args nowhere near `ndists`

CoryMcCartan commented on R/redist_map.R:
```diff
@@ -9,8 +9,10 @@
-new_redist_map <- function(data, adj, ndists, pop_bounds, pop_col = "pop",
-                           adj_col = "adj", add_adj = TRUE, existing_col = NULL) {
+new_redist_map <- function(
+        data, adj, ndists, total_seats, district_seat_sizes,
```
@philipwosull 
rename to `nseats` for consistency and `seats_range`
let's default `nseats=ndists` and `seats_range=c(1, 1)` or `1` or whatever the format is

CoryMcCartan replied:
Ah, sorry, this is the internal constructor. The comment about defaults applies to `redist_map` below

CoryMcCartan commented on R/redist_map.R:
```diff
@@ -223,6 +263,35 @@ redist_map <- function(..., existing_plan = NULL, pop_tol = NULL,
+    }else{
+        # check the district sizes are numbers
+        if(!is.numeric(district_seat_sizes)){
+            cli::cli_abort("{.arg district_seat_sizes} must be integers.")
```
can remove `cli::`

CoryMcCartan commented on R/redist_ms.R:
```diff
@@ -63,13 +69,19 @@
 #' @param init_name a name for the initial plan, or \code{FALSE} to not include
 #' the initial plan in the output.  Defaults to the column name of the
 #' existing plan, or "\code{<init>}" if the initial plan is sampled.
-#' @param silly_adj_fix Heuristic for fixing weird inputs.
```
@christopherkenny was this argument you !!?

CoryMcCartan commented on R/diagnostics.R:
```diff
@@ -81,91 +88,274 @@ summary.redist_plans <- function(object, district = 1L, all_runs = TRUE, vi_max
+        if(1 <= district && district <= n_distr){
+            rhat_df <- object |>
+                filter(!is.na(chain) & district == !!district) |>
+                tidyr::pivot_longer(
```
Also, any variables you use inside `dplyr` verbs unqouted need to be escaped with `data$` or `utils::globalVariables()`

CoryMcCartan commented on R/redist_smc.R:
```diff
@@ -123,266 +214,780 @@
-    # Other constraints
-    if (!inherits(constraints, "redist_constr")) {
-        constraints <- new_redist_constr(eval_tidy(enquo(constraints), map))
+    if (!assertthat::is.scalar(compactness) || compactness < 0)
```
can't use assertthat

philipwosull replied:
Removed all assertthat calls from `redist_smc`

CoryMcCartan commented on R/redist_smc.R:
```diff
@@ -123,266 +214,780 @@
             cli::cli_process_done()
-            cli::cli_process_done()
         }
+        if(DEBUG_MODE) print("Done with c++ Code!")
```
remove all of these `DEBUG_MODE` things; and the global constant

CoryMcCartan commented on R/redist_smc.R:
```diff
@@ -123,266 +214,780 @@
+            # also add NA
+            sd_lp <- c(sd_lp, NA)
+
+            #TODO probably need to adjust the rest of these as well
```
resolve

philipwosull replied:
removed line

CoryMcCartan commented on R/region_counting.R:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    # create hash map
+    region_count_map <- hash::hash()
```
no `hash`

use `rlang::hash` if you need something like this

CoryMcCartan commented on R/region_counting.R:
How are these functions used?

CoryMcCartan reviewed (changes requested):
@philipwosull a first high-level review. I only went through `R/`; we may deal with C++ later.

Overall comments:

- The code style is not good, especially the spacing around `if`, `function`, etc.  Please standardize to match the other code and the tidyverse style guide. @christopherkenny should we just start using `air` on this project?

- You need to pass `R CMD CHECK` with no errors, warnings, or notes

- Your recent changes also created merge conflicts; please resolve

Ping when these are all addressed and Chris and I can do a second pass

CoryMcCartan commented:
add to `NEWS.md` under `redist 5.0` your major changes, esp. including breaking changes

philipwosull committed:
adding some of the new updates under 5.0

philipwosull committed:
more tweaks to news and readme

philipwosull committed:
addressing comments on plans_helpers

philipwosull committed:
updating main readme

philipwosull commented on .Rbuildignore:
Not sure. Reverted

philipwosull committed:
reverting changes to dot files

philipwosull commented on R/redist_smc.R:
```diff
@@ -123,266 +214,780 @@
-#' # One run with multiple cores
-#' redist_smc(fl_map, 1000, ncores = 2)
+#' # One run with multiple processes
+#' redist_smc(fl_map, 1000, nproc = 2)
```
`control` argument. Controls how many independent instances of R to create.

philipwosull committed:
starting to work on short explainer for c++ code

philipwosull committed:
revising redist smc

philipwosull committed:
switching naming conventions. total_seats to nseats, nseats to seats, district_seat_sizes to seats_range, and init_nseats to init_seats

philipwosull committed:
switched shortburst mergesplit backend to forest space

philipwosull committed:
more renaming of region_sizes to seats

philipwosull commented on R/redist_plans.R:
```diff
@@ -128,6 +198,7 @@ reconstruct.redist_plans <- function(data, old) {
 #' @param map a \code{\link{redist_map}} object
 #' @param algorithm the algorithm used to generate the plans (usually "smc" or "mcmc")
 #' @param wgt the weights to use, if any.
+#' @param region_nseats The number of seats in each region (if they are not all the same).
```
changed to `seats`

philipwosull commented on R/redist_plans.R:
```diff
@@ -185,6 +263,46 @@ set_plan_matrix <- function(x, mat) {
+#' @return matrix
+#' @concept analyze
+#' @export
+get_nseats_matrix <- function(x) {
```
yes, changed to `get_seats_matrix`

philipwosull committed:
cleaning up redist alg helpers

philipwosull committed:
cleaning up redist alg helpers

philipwosull committed:
removing cli prefix before cli_warn and cli_abort

philipwosull commented on R/redist_ms.R:
```diff
@@ -88,161 +100,393 @@
+#' sampled_basic <- redist_mergesplit(fl_map, 10000, chains = 10)
 #'
-#' sampled_constr <- redist_mergesplit(fl_map, 10000, constraints = list(
+#' sampled_constr <- redist_mergesplit(fl_map, 10000, chains = 10,
```
@CoryMcCartan Setting `ncores = 2` means only 2 processes will be used to run 10 chains. Updated to add `ncores = 2`

philipwosull commented on R/redist_ms.R:
```diff
@@ -88,161 +100,393 @@
+    sampling_space = c("graph_plan", "spanning_forest", "linking_edge"),
+    split_method = NULL, split_params = NULL,
+    merge_prob_type = "uniform", init_nseats = NULL,
+    ncores = NULL,
```
For `ncores` it controls the number of parallel processes spawned for the doPar for loop

philipwosull committed:
refactored summary function

philipwosull committed:
now pass district pops out from redist smc and ms to avoid needing to recompute

philipwosull commented on R/diagnostics.R:
```diff
@@ -81,91 +88,274 @@ summary.redist_plans <- function(object, district = 1L, all_runs = TRUE, vi_max
+    # print algorithm specific parameters
+    if (algo %in% c(SMC_ALG_TYPE, MS_SMC_ALG_TYPE)) {
+        cli::cli_text("SMC Parameters: {.arg weight_type}={format(all_run_info[[1]]$weight_type, digits=2)}
+        \u2022 {.arg seq_alpha}={format(all_diagn[[1]]$seq_alpha, digits=2)}
```
Updated to bulleted list

philipwosull commented on R/diagnostics.R:
```diff
@@ -57,17 +60,21 @@
 #' @md
 #' @export
-summary.redist_plans <- function(object, district = 1L, all_runs = TRUE, vi_max = 100, ...) {
+summary.redist_plans <- function(
```
Seperated creating diagnostic dataframe for SMC and mergesplit steps into their own functions.

philipwosull commented on R/diagnostics.R:
```diff
@@ -81,91 +88,274 @@ summary.redist_plans <- function(object, district = 1L, all_runs = TRUE, vi_max
+                summarise(rhat = diag_rhat(value, chain, split=split_rhat), .groups = "drop")
+        }else{
+            # else compute rhats using existing district number
+            rhat_df <- object |>
```
Wrote helper function for computing rhats.

philipwosull commented on R/diagnostics.R:
```diff
@@ -320,3 +463,106 @@ diag_rhat <- function(x, grp, split = FALSE) {
+#'
+#' @returns indices of k-step ancestors for particles at iteration `start_col`
+#' @md
+#' @export
```
No longer exported.

philipwosull commented on R/redist_ms.R:
```diff
@@ -88,161 +100,393 @@
     }
 
-    out
+    dplyr::relocate(out, chain, .after = "draw")
```
@CoryMcCartan What order should they be in? Right now it returns `draw, chain, district, total_pop, mcmc_accept`. In constrast `redist_smc` returns `draw, district, total_pop, chain`

philipwosull committed:
renaming score_districts_only to only_districts for constraints

philipwosull committed:
added thresh to constraint R code, still need to add to c++

philipwosull committed:
more minor tweaks

philipwosull committed:
moved some constraint functions to map calc

christopherkenny reviewed (changes requested):
So the good thing is that I don't have huge comments, but I do have many, many, many small comments. 

There is a ton of work to do here, especially to conform with R best practices. We want the code to work well for everyone, both existing and new, which requires carefully removing options across at least 2 versions.

On a less specific note, this PR is hard to review because there are 100s of lines that are edited because of whitespace changes and not because of meaningful changes. If you can review the substantive changes, then Cory and I can both re-review it.

philipwosull commented on R/redist_ms.R:
```diff
@@ -90,159 +100,481 @@
+  districting_scheme <- map_params$districting_scheme
+
+    # check that no district is the sum of two others
+    for (d_size1 in seats_range) {
```
Unfortunately yes. Those examples don't pass validation and we don't know why yet.

philipwosull commented on explore/ms_smc.R:
`adapt_k_thresh` is only relevant if the sampling space is graph space. For the other sampling spaces they have different parameters and I think it would make things very unwieldy to make each possible splitting parameter a separate argument.

philipwosull commented on R/redist_shortburst.R:
```diff
@@ -175,20 +165,46 @@ redist_shortburst <- function(map, score_fn = NULL, stop_at = NULL,
+            # just assume always SMD
+            init_sizes <- matrix(1L, ncol = 1, nrow = ndists)
+
+            ms_plans(
```
If I have time I will make the sampling space an input option like `redist_smc` or `redist_mergesplit` but for now this is a much better option than graph space (generally higher acceptance rates since it doesn't need to do the k estimation thing).

philipwosull commented on README.md:
```diff
@@ -86,8 +93,9 @@ data(iowa)
+#> Starting Chain 1
 #> SEQUENTIAL MONTE CARLO
-#> Sampling 500 99-unit maps with 4 districts and population between 760,827 and 762,350.
+#> Using Graph Sampling space to sample 500 99-unit maps with 4 districts and population between 760827 and 762350.
```
Not sure what needs to be done in c++ for that. As far as I can tell I still have the same `set_precision` and `        Rcout.imbue(std::locale::classic());` code

philipwosull commented on R/redist_map.R:
```diff
@@ -238,13 +323,46 @@ redist_map <- function(..., existing_plan = NULL, pop_tol = NULL,
+                nseats
+            )
+            # flatten to vector
+            dim(existing_col_seats) <- NULL
```
For small matrices like this though I can change it.

philipwosull committed:
changes to linking edge multigraph tau

philipwosull commented on R/redist_alg_helpers.R:
```diff
@@ -0,0 +1,502 @@
+  # first check its a valid sampling space
+  sampling_space <- rlang::arg_match(
+    sampling_space,
+    values = VALID_SAMPLING_SPACES
```
@christopherkenny I don't follow? What do you mean by this?

philipwosull commented on R/redist_ms.R:
```diff
@@ -90,159 +100,481 @@
+
+
+  # subtract 1 to make it 0 indexed
+  init_plan <- init_plan - 1
```
We need to decide this package-wide then. The convention for my c++ code is all plan related things are passed in zero-indexed. If its consistent with older code I'm fine to always assume everything passed in from R is 1-indexed it just needs to be consistent.

philipwosull commented on R/redist_ms.R:
```diff
@@ -90,159 +100,481 @@
+    on.exit(stopCluster(cl))
+
+    # this makes it avoid printing the loading required package message each time
+    parallel::clusterEvalQ(cl, {
```
Removed

philipwosull commented on R/redist_smc.R:
```diff
@@ -123,277 +209,934 @@
+      )
+    }
+    # this makes it avoid printing the loading required package message each time
+    parallel::clusterEvalQ(cl, {
```
Before if you had more runs than processes after the first one is complete it would just be silent. This ensures one of them is always printing if `silent=F`.

For example, imagine I have 5 processes and 30 runs. Without this line it will print information for a single run and then stop. This ensures one of the processes can always be non-silent if that is desired.

philipwosull commented on R/redist_smc.R:
```diff
@@ -123,277 +209,934 @@
+    }
+    # this makes it avoid printing the loading required package message each time
+    parallel::clusterEvalQ(cl, {
+      suppressPackageStartupMessages(library(foreach))
```
removed.

philipwosull committed:
in the middle of fixing linking edge with counties on

philipwosull committed:
hierarchical linking edge count now passes some tests

philipwosull committed:
linking edge with counties on now passes first round of tests. Method for computing for merged plans is still very memory inefficient though

philipwosull committed:
linking edge now passes all hierarchical validations

philipwosull committed:
resolving some merge conflicts

philipwosull committed:
all cli lib calls now prefixed with cli colon colon

philipwosull committed:
slightly sped up hierarchical linking edge calculation

philipwosull committed:
more efficiency tweaks to compute log linking edge

philipwosull committed:
for optimal weights moved linking edge computation from weights

philipwosull committed:
flipped linking edge term ratio in optimal weights. fixed

philipwosull committed:
added and tested faster method to convert log linking edge for merged plans

philipwosull committed:
now only compute linking edge selection probability when needed, instead of preemptively

philipwosull committed:
fixed issue with district populations not being resampled correctly

philipwosull committed:
when inferring seat sizes only warns about not-tight bounds once

philipwosull committed:
added mergesplit parallel back with deprecation warning

philipwosull commented on DESCRIPTION:
```diff
@@ -1,10 +1,11 @@
 Package: redist
-Version: 4.2.0.9000
+Version: 5.0.0.1
```
Changed to 5.0.0

philipwosull commented on NAMESPACE:
```diff
@@ -175,9 +177,7 @@ export(redist_flip_anneal)
 export(redist_map)
 export(redist_mcmc_ci)
 export(redist_mergesplit)
-export(redist_mergesplit_parallel)
```
Put `redist_mergesplit_parallel` back and made it just throw an abort message to use redist_mergesplit

philipwosull committed:
max_dev now works for multimember plans

philipwosull committed:
more code tweaks

philipwosull committed:
more mostly style changes for pr

CoryMcCartan commented on R/diagnostics.R:
```diff
@@ -81,99 +91,247 @@ summary.redist_plans <- function(object, district = 1L, all_runs = TRUE, vi_max
+        rhats_p <- vapply(max_rhats, function(x){
+            ifelse(x <= q99_rhat_thresh, sprintf('%.3f', x), paste0('\U274C', round(x, 3)))
+        }, FUN.VALUE = character(1))
+        print(noquote(rhats_p))
```
`print` formats the numbers nicely and spaces them out

until recently you couldn't tell if something was `cli`'d because there was no blue line in the console. now there is so we should figure out some way to consistently format everything.  Ideally `cat` rather than `cli` I guess

philipwosull committed:
added thresholding in constraints and new test for custom plan based constraints

philipwosull committed:
adding some missing man pages

philipwosull committed:
made summary backwards compatible with previous redist plans

philipwosull committed:
moved constants to package file

christopherkenny reviewed (commented):
This is much closer. I've added more comments and some existing ones were not yet resolved. Thanks Philip.

philipwosull committed:
modified constraints cpp code to reduce memory usage

philipwosull committed:
thresholding now turns constraints into indicators

CoryMcCartan reviewed (changes requested):
@philipwosull when you run the formatter you have to match our existing formatting! 4-space indents among other things. That created a lot of diff noise that makes it hard to tell what substantively changed.

what is the status of `devtools::check()` at this point?

CoryMcCartan commented:
@philipwosull when you get a chance, merge in `main` again which will hopefully get `R CMD CHECK` automatically running on here every time you push

philipwosull committed:
added missing array header

philipwosull committed:
can now exactly specify what number of regions plan constraints apply to

philipwosull committed:
fixed problem with non-unique thread ids

philipwosull committed:
fixed remaining thread id assignment issues. Now ensures thread ids are generated fresh for each function call

philipwosull committed:
fixed redist parity bug with MMD

philipwosull committed:
renaming some mergesplit parameters

philipwosull committed:
removing stale debugging statement

philipwosull committed:
forgot to rename some ms move to step in diagnostics

philipwosull committed:
updated diagnostics to return rhats and diagnostics df

philipwosull committed:
fixed polsby popper calculations

philipwosull committed:
fixed minor bug with competition constraint c++ code

philipwosull committed:
updating some doc files

philipwosull committed:
minor verbose output printing tweaks

philipwosull committed:
added min_frac thresholding for multiple groups and changed default cores to 1 for redist smc and ms

philipwosull committed:
added min_frac thresholding for multiple groups and changed default cores to 1 for redist smc and ms

philipwosull committed:
adding more doc files

philipwosull committed:
for smc can now specify mergesplit for last n steps or every n steps

philipwosull committed:
added plan total splits constraint

philipwosull committed:
fixed issue with redist ms test

philipwosull committed:
adding partial caching for compactness calculations

philipwosull committed:
new much faster tau calculations

philipwosull committed:
removed redundant print statement

philipwosull committed:
fixed an issue with cache updating between smc rounds

philipwosull committed:
switched linking edge taus to eigen as well:

philipwosull committed:
Merge pull request #4 from philipwosull/sparsechol

Sparsechol

CoryMcCartan committed:
Update pkgdown workflow triggers

Remove push and pull_request triggers from pkgdown workflow.

CoryMcCartan committed:
Update check-standard.yaml to simplify branch rules

Remove pull request branch specification for workflow.

CoryMcCartan committed:
Add workflow_dispatch trigger to check-standard.yaml

philipwosull committed:
fixing issue with missing functions in RcppExports

philipwosull committed:
fixing issue with subsetting when chain values are NA

philipwosull committed:
improving computational stability for plan based constraints

philipwosull committed:
adding max_split_tries parameter

CoryMcCartan committed:
weaken impossibility check

philipwosull committed:
improving some linking edge error diagnostics

philipwosull committed:
more diagnostic printing detail

philipwosull committed:
rbind for 4.0 and 5.0 SMD plans should work now

philipwosull committed:
shortburst now works for multi member district plans

philipwosull committed:
updating shortburst for mmd init plan

CoryMcCartan committed:
copy config files from dev

CoryMcCartan committed:
run formatters

CoryMcCartan committed:
resolve merge conflicts

CoryMcCartan committed:
Get tests running, integrate cycle walk, fix lints

CoryMcCartan committed:
reorg ignores

CoryMcCartan committed:
base pipe, clean tests, fix check?

CoryMcCartan committed:
cpp format and try to fix windows check

CoryMcCartan committed:
printfs for diagnosing windows CI

CoryMcCartan committed:
trace v2

CoryMcCartan committed:
trace 3

CoryMcCartan committed:
trace 4

CoryMcCartan committed:
trace 5

CoryMcCartan committed:
trace 6

CoryMcCartan committed:
revert diagnostics & remove threading from MS setup (should fix for now?)

CoryMcCartan committed:
refactor thread_local statics

philipwosull committed:
ensuring shortburst status quo scorer properly labels plans

philipwosull committed:
adding issue with backwards compatibility of adding reference plans

philipwosull committed:
removing all calls to arma log symmetrical determinant (replaced with eigen)

philipwosull committed:
allowing mmd sampling when one district size can be written as sum of two others with a warning that convergence to target is guaranteed at the moment

philipwosull committed:
removing erroneous warnings about certain mmd size configurations

philipwosull committed:
adding some more granular time tracking of smc code

philipwosull committed:
making debugging if statements constexpr and making redist ms use less memory for only a single chain run

philipwosull committed:
adding toggable dev flag for granular time tracking

philipwosull committed:
removing some extraneous at instead of bracket access

philipwosull committed:
updating random func to be pass by ref

philipwosull committed:
added packed forest but didn't remove old type

philipwosull committed:
still occasionally throwing up some kind of rare bug

philipwosull committed:
adding more checks in the code

philipwosull committed:
more error catching added

philipwosull committed:
even more checks

philipwosull committed:
more error checks again

philipwosull committed:
more checks

philipwosull committed:
removing smc base and making utils file

philipwosull committed:
made debug checks more robust

philipwosull committed:
adding more optional checks

philipwosull committed:
even more checks added

philipwosull committed:
last few commits before removing old forest graph code

philipwosull committed:
removing unnesscary checks and putting the others behind compile flags

philipwosull committed:
added proper dummy county sampling

philipwosull committed:
removed vertex forest as an option entirely

philipwosull committed:
switching parent tracking from atomics to one vector per thread (more memory usage in exchange for less thread conflict)

philipwosull committed:
removed old graph tree code

philipwosull committed:
slightly faster forest updating now

philipwosull committed:
makes vectors appended instead of inserted in find joined tree edges

philipwosull committed:
switched dummy county trees to BFS to encourage bushier trees

philipwosull committed:
brought back compute retroactive split prob with vertex forest for forest space sampling

philipwosull committed:
removed duplicate constraint code and wrote helpers for R processing

philipwosull committed:
speeding up log eff boundary computation for forest space by only copying used parts of forest into graoph vertex

philipwosull committed:
turning off granular tracking for main branch

philipwosull committed:
removing debugging check from smc

philipwosull committed:
Merge pull request #7 from philipwosull/packedforests

Packedforests

philipwosull committed:
now draws fake dummy trees for graph space wilson again

philipwosull committed:
added timing to verbose printouts and consolidated USTSampler class into Wilson files

philipwosull committed:
removed fake dummy trees entirely

philipwosull committed:
more tweaks to Wilson code and USTSampler calls

philipwosull committed:
refactoring to try to reduce compile time and size

philipwosull committed:
code refactoring which reduces compile time by 40percent on local machine

Commit:
forgot some missing headers

philipwosull committed:
fixing overflow error in flatgraph

philipwosull committed:
switching to FlatGraph for forest space eff boundary calculations

philipwosull committed:
switching back to vector trees for Wilson trees

philipwosull committed:
slight tweaks to wilson code

philipwosull committed:
turning off granular tracking by default

philipwosull committed:
Begin hierarchical Wilson overhaul

philipwosull committed:
starting overhaul

philipwosull committed:
starting to refactor Wilson code

philipwosull committed:
switching to new admin walk until function and more streamlining

philipwosull committed:
adding backfilling deterministic trees

philipwosull committed:
now tracking wilson backfilling time as seperate

philipwosull committed:
removing unncessary fill in forest weights

philipwosull committed:
Merge pull request #8 from philipwosull/hierwilson

Hierwilson additions

philipwosull committed:
Disable granular tracking for performance analysis

Changed granular tracking performance configuration to false for main branch use

philipwosull committed:
small tweaks to reduce mem usage

philipwosull committed:
Include memory header in splitting_schedule_types.cpp

Added memory header for smart pointer usage.

philipwosull committed:
Add memory header inclusion to splitting_schedule_types.h

philipwosull committed:
starting final eff boundary computation upgrade to linear time computation

philipwosull committed:
adding missing mem includes

philipwosull committed:
added faster forest weight calculation for simple population based splitters but left in redudancy checks for now

philipwosull committed:
fixed an indexing bug in the district size checker constraint

philipwosull committed:
removing redundant forest checks

philipwosull committed:
Merge pull request #9 from philipwosull/hierwilson

Faster Forest Weights
